### DISCUSSION TITLE
How far away from the TOP500 supercomputers ?

### DISCUSSION THREAD

**Discussion by SuperUserNameMan** [dbuezas/lgt8fx#102] [Ideas]
Hello,

This afternoon, I was wondering how many FLOPS could perform the LGT8F32.

According to this [LINPACK benchmark port to Arduino](https://gist.github.com/projectgus/8279947), here are the results :

| optimization level | Mega FLOPS @ 32MHz | FLOPS / MHz |
|:-----------------------:|:--------------------------------:|:-------------------:|
| Os                        | ~0.205                            |  ~6406             |
| O0                        | ~0.160                            |  ~5000             |
| O1                        | ~0.204                            |  ~6375             |
| O2                        | ~0.205                            |  ~6406             |
| O3                        | ~0.220                            |  ~6875             |
| Ofast                    | ~0.226                            |  ~7062             |

Note : To change the optimization level, you can add `#pragma GCC optimize("Ox")` at the beginning of the benchmark source code.

Conclusion : to enter the TOP500 from the bottom, we'd need at least 6 Billions LGT8F328p.

---

I also quickly ported a C version of the [FBENCH](https://gist.github.com/SuperUserNameMan/0a115805c2e7169c39806d35193b8798) benchmark from [here](https://www.fourmilab.ch/fbench/fbench.html).

| optimization level | 1000 iterations @ 32MHz (lower better) |
|:-----------------------:|:----------------------------------:|
| Os                        | 4.670 s                             |
| O0                        | 5.105 s                             |
| O1                        | 4.657 s                             |
| O2                        | 4.651 s                             |
| O3                        | 4.651 s                             |
| Ofast                    | 4.605 s                             |

It puts the LGT8F328p @ 32MHz at the same level than a Sun 3/260 ( Motorola 68020 CPU @ 25 Mhz + 20 MHz FPU ) from the 80's.

( a i386 @25MHz + 387 math coprocessor score is 3.0s, while a Pentium 4 @ 3.4 GHz score is 0.00862s ).

**Comment by SuperUserNameMan** [comment]
# LGT8F328p vs ATmega328p

When compared to a real ATmega328p, the LGT8F328p gives around 22% more performance in the Arduino-LINPACK benchmark, and is 1.17 times faster in the fbench test.

That would be amazing results (if better alternatives like STM32 and ESP did not exist, and if we were still in the 80's ....)

On 
---

**"Arduino-LINPACK" benchmark :** results in FLOPS / MHz at 16MHz (higher better)
| Optimization Level | LGT     | ATmega |
|:-------------------------:|:--------:|:-----------:|
| O3                          | ~6860 | ~5619 |
| Ofast                      | ~7031 | ~5742 |

---

**fbench @ 16MHz :** results in seconds (lower better)
| Optimization Level | LGT      | ATmega |
|:-------------------------:|:---------:|:-----------:|
| O3                          |  9.301 s  | 10.894 s  |
| Ofast                      |  9.209 s | 10.774 s |

**Comment by dbuezas** [comment]
This is fascinating, I always wondered about this. Once I benchmarked the attiny85 against commodore 64 basic (inverting matrices)

The attiny85@16Mhz won. With a caveat: the float precision in 8 bit arduinos quite low, 4 bytes per float. C64 basic uses 5 bytes, and 80s PCs... I guess it depends on the benchmark?

In any case, the lgt extra speed is awesome. I made a pitch shifter outputting through the DAC (opamp buffered) and the board has no issues pulling 300kSamples/s while outputting at 22kHz. Also sensor fusion is very snappy on these, I reached 1kHz update rate fusing 9 axes (mag, acc and gyros). The limitation was i2c, but there's an asynchronous lib for i2c that pretty much solves it (processing was done while communication was going)

Thanks for this benchmark, it is so random I love it!
